07_StagingAndMultipleQueues/main.cpp

```diff
@@ -2,78 +2,580 @@
+				logFailAndTerminate("Couldn't create descriptor.");
+			imgInfo.info.image = { .sampler = sampler, .imageLayout = IImage::LAYOUT::GENERAL };
+
+			IGPUDescriptorSet::SDescriptorInfo bufInfo;
+			bufInfo.desc = smart_refctd_ptr(histogramBuffer);
+			bufInfo.info.buffer = { .offset = 0u, .size = histogramBuffer->getSize() };
+
+			IGPUDescriptorSet::SWriteDescriptorSet writes[2] = {
+				{.dstSet = descSets[resourceIdx].get(), .binding = 0, .arrayElement = 0, .count = 1, .info = &imgInfo },
+				{.dstSet = descSets[resourceIdx].get(), .binding = 1, .arrayElement = 0, .count = 1, .info = &bufInfo }
+			};
+			m_device->updateDescriptorSets(2, writes, 0u, nullptr);
+
+			computeQueue->startCapture();
+			cmdBuff->begin(IGPUCommandBuffer::USAGE::NONE);
+			cmdBuff->beginDebugMarker("My Compute Dispatch", core::vectorSIMDf(0, 1, 0, 1));
+			cmdBuff->bindComputePipeline(pipeline.get());
+
+			cmdBuff->bindDescriptorSets(nbl::asset::EPBP_COMPUTE, pplnLayout.get(), 0, 1, &descSets[resourceIdx].get());
+
+			const auto imageExtent = images[imageToProcessId]->getCreationParameters().extent;
+			const uint32_t wgCntX = imageExtent.width / WorkgroupSizeX;
+			const uint32_t wgCntY = imageExtent.height / WorkgroupSizeY;
+
+			PushConstants constants;
+			constants.histogramBufferOffset = HISTOGRAM_SIZE * resourceIdx;
+
+			cmdBuff->pushConstants(pplnLayout.get(), IShader::E_SHADER_STAGE::ESS_COMPUTE, 0u, sizeof(PushConstants), &constants);
+			cmdBuff->dispatch(wgCntX, wgCntY, 1);
+
+			cmdBuff->endDebugMarker();
+			cmdBuff->end();
+
+			IQueue::SSubmitInfo submitInfo[1];
+			IQueue::SSubmitInfo::SCommandBufferInfo cmdBuffSubmitInfo[] = { {cmdBuff.get()} };
+			IQueue::SSubmitInfo::SSemaphoreInfo signalSemaphoreSubmitInfo[] = { {.semaphore = m_imagesProcessedSemaphore.get(), .value = imageToProcessId+1, .stageMask = PIPELINE_STAGE_FLAGS::ALL_COMMANDS_BITS } };
+			submitInfo[0].commandBuffers = cmdBuffSubmitInfo;
+			submitInfo[0].signalSemaphores = signalSemaphoreSubmitInfo;
+
+			waitForResourceAvailability(m_histogramSavedSeamphore.get(), imageToProcessId);
+			computeQueue->submit(submitInfo);
+			computeQueue->endCapture();
+			std::string msg = std::string("Image nr ") + std::to_string(imageToProcessId) + " processed. Resource idx: " + std::to_string(resourceIdx);
+			m_logger->log(msg);
+		}
+	}
 
+	void saveHistograms()
+	{
+		std::array<std::ofstream, IMAGE_CNT> files;
+		for (auto& file : files)
+		{
+			static uint32_t i = 0u;
+			file.open("histogram_" + std::to_string(i++) + ".csv", std::ios::out | std::ios::trunc);
+		}
+
+		for (uint32_t imageHistogramIdx = 0; imageHistogramIdx < IMAGE_CNT; ++imageHistogramIdx)
+		{
+			waitForPreviousStep(m_imagesProcessedSemaphore.get(), imageHistogramIdx + 1);
+
+			const uint32_t resourceIdx = imageHistogramIdx % FRAMES_IN_FLIGHT;
+			uint32_t* histogramBuff = m_histogramBufferMemPtrs[resourceIdx];
+
+			waitForResourceAvailability(m_histogramSavedSeamphore.get(), imageHistogramIdx);
```

Comment:
this is redundant, because its the histogram computing dispatch (image processing) needs to wait for the histogram output range to stop being used

Reply:
and because you're wiping the histogram it needs to wait on `m_histogramSavedSemaphore` before the dispatch